### ISSUE TITLE
StackOverflowError on TypeArguments.apply(typeDef) for MultiMaps & other non-trivial generics

### ISSUE THREAD

**Issue by xRodney** [sundrio/sundrio#333]
Hello,

I ran into this error while working on https://github.com/fabric8io/kubernetes-client/issues/4357. As written there, sundrio is looping infinitely, on some more complicated generic classes.

An example of such class is a multi map:
```
public class MultivaluedHashMap<K, V> extends HashMap<K, List<V>> { /* ...  */ }
```

In kubernetes-client, the code involved in the loop is in part custom to that project, which is why at first I thought the error is there. However, then I found almost identical code is also present in sudrio project. 
In both cases, the code is responsible for "inflating" TypeDeft to include also members from their superclasses. 
In sundrio, this logic is implemented in an util `TypeArguments`
So I wondered if it suffers from the same bug and sure enough, it stack overflows as well.

Here's a minimal failing example:
```
package io.sundr.model.utils;

import org.junit.Test;

import io.sundr.model.ClassRef;
import io.sundr.model.Kind;
import io.sundr.model.TypeDef;
import io.sundr.model.TypeDefBuilder;
import io.sundr.model.functions.GetDefinition;
import io.sundr.model.repo.DefinitionRepository;

import static io.sundr.model.utils.Collections.K;
import static io.sundr.model.utils.Collections.V;

public class MultiMapTest {
  public static final TypeDef MULTI_HASH_MAP = new TypeDefBuilder(TypeDef.forName("com.example.MultiHashMap"))
          .withKind(Kind.CLASS)
          .withParameters(K, V)
          .withExtendsList(Collections.HASH_MAP.toReference(K.toReference(), Collections.LIST.toReference(V.toReference())))
          .build();

  @Test
  public void testName() {
    DefinitionRepository.getRepository().register(Collections.MAP);
    DefinitionRepository.getRepository().register(Collections.HASH_MAP);
    DefinitionRepository.getRepository().register(Collections.LIST);
    DefinitionRepository.getRepository().register(Types.STRING);
    DefinitionRepository.getRepository().register(Types.INT);
    DefinitionRepository.getRepository().register(MULTI_HASH_MAP);

    ClassRef classRef = MULTI_HASH_MAP.toReference(Types.STRING_REF, Types.INT_REF);
    TypeDef typeDef = GetDefinition.of(classRef);

    System.out.println(classRef.render());
    System.out.println(typeDef.render());

    TypeArguments.apply(typeDef);    //////// StackOverflowError
  }
}
```

In `TypeArguments` util, there are three visitors:
```
return new TypeDefBuilder(definition)
            .accept(
                    new ApplyTypeParamMappingToProperty(mappings, ORIGINAL_TYPE_PARAMETER),
                    new ApplyTypeParamMappingToMethod(mappings, ORIGINAL_TYPE_PARAMETER),
                    new ApplyTypeParamMappingToTypeArguments(mappings))    /// <---- this one causes the stack overflow
            .build();
```

For all three, the `mappings` is a map from generic parameter to its value. For example for `class MyMap implements Map<String,Integer>` and `interface Map<K,V>` the mapping might be `{K -> String, V -> Integer}`. 
In our case, the mapping happens to be `{K -> String, V -> List<V>}`. I *think* that the `V` is being inflated repeatedly to `List<List<List<...>>>` until memory runs out.

In https://github.com/fabric8io/kubernetes-client/issues/4357 I propose a fix of marking the `V` unresolvable with an attribute to terminate the loop. It's in this commit https://github.com/fabric8io/kubernetes-client/commit/d62cba1f842b1603409b1f915b238d0bd6480f48 and it seems to resolve the problem. 

(Note that the patch is against the copy/pasta of `TypeArguments`, but I hope it illustrates the proposed concept well enough).

I volunteer to port it here, with your permission and if there isn't a better way of fixing it.

**Comment by iocanel** [comment]
@xRodney: You're spot on! The `V -> List<V>` is causing the whole thing and it happens because `HashMap` internally uses `K` and `V` as parameters and the `MultiHashMap` maps `V` to `List<V>`.

I am looking into ways of breaking this loop.

**Comment by iocanel** [comment]
Fixed in #334

**Comment by xRodney** [comment]
Hello @iocanel, 

I am afraid replacing `V -> List<V>` with `V -> List<?>` is not a valid substitution. While it eliminates the looping, it also pollutes the methods and properties coming from supertype. For example, the `crd-generator` iterates over properties from the type and its supertypes to generate the schema. It also considers getters/setter to load additional info.  So it's important that the type is as accurate as possible.

Please see [this failing test](https://github.com/sundrio/sundrio/commit/64019bfe7d8dc6b729d9e41bb77291d5fdf2343e).

I am preparing PR with a fix, I'd be happy for a review.
